File: 00204.py
# https://leetcode.com/problems/count-primes/

class Solution:
    def countPrimes(self, n: int) -> int:
        if n < 3:
            return 0
        elif n == 3:
            return 1
        
        primes = []
        if n%2 == 0:
            primes = [False]*2+[True]*2 + [False, True]*(n//2-2)
        else:
            primes = [False]*2+[True]*2 + [False, True]*(n//2-2) + [False]

        for i in range(3, int(sqrt(n))+1,2):
            if primes[i]:
                for j in range(i*i, n, i):
                    primes[j] = False

        return sum(primes)

File: 00204.c
// https://leetcode.com/problems/count-primes/

int countPrimes(int n) {
    if(n < 3)
        return 0;

    int *primes = (int *) malloc((n+1)*sizeof(int));
    // presumir que todos ímpares são primos
    for(int i = 3; i <= n; ++i)
        primes[i] = 1;
    
    // invalidar pares
    for(int i = 4; i < n; i += 2) primes[i] = 0;
    
    int max = sqrt(n);
    for(int i = 3; i <= max; i += 2){
        if(primes[i])
            for(int j = i*i; j < n; j += i)
                primes[j] = 0;
    }
    
    // contar primos:
    int amount = 1; // 2
    for(int i = 3; i < n; i += 2)
        if(primes[i]) 
            ++amount;
    
    free(primes);
    return amount;
}
